### PR TITLE
doc: Add non-style-related development guidelines

### PR DESCRIPTION
**Pull request by laanwj**
I've collected these over time, mostly adding notes after troubleshooting obscure bugs. As I hope to get the community more involved in the whole process, I think it is useful to add to the developer-notes.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Nice. I think this is very helpful for new devs (and even for old hands).

I would say we should take these new point directly at the 2nd position above of "doxygen comments".

**Comment by paveljanik** [comment]
These are gems! Thank you for collecting all these notes. They help a lot!

**Comment by petertodd** [comment]
Awesome, thanks!

@paveljanik  I'd prefer dots at the end of sentences too.

ACK

**Comment by gmaxwell** [comment]
ACK (with or without my suggestions), this is all good stuff.

**Comment by sipa** [comment]
Very ACK.

**Comment by gavinandresen** [comment]
ACK. Very nice!

File: doc/developer-notes.md [hunk 1]
```diff
@@ -204,3 +204,172 @@ If a set of tools is used by the build system or scripts the repository (for
...
+
+General Bitcoin Core
+----------------------
+
+- New features should be exposed on RPC first, then can be made available in the GUI
+
+  - *Rationale*: RPC allows for better automatic testing. The test suite for
+    the GUI is very limited
+
+- Make sure pulls pass Travis CI before merging
```
Comment (omefire) [inline doc/developer-notes.md]: very minor nitpick: instead 'pulls', should we make it 'pull requests' ?

File: doc/developer-notes.md [hunk 2]
```diff
@@ -204,3 +204,172 @@ If a set of tools is used by the build system or scripts the repository (for
...
+     on the master branch. Otherwise all new pull requests will start failing the tests, resulting in
+     confusion and mayhem
+ 
+  - *Explanation*: If the test suite is to be updated for a change, this has to
+    be done first 
+
+Wallet
+-------
+
+- Make sure that that no crashes happen with run-time option `-disablewallet`.
```
Comment (omefire) [inline doc/developer-notes.md]: nitpick: 'that' appears twice.

**Comment by omefire** [comment]
Thanks, these are helpful, especially for folks like me, ramping up on the codebase and planning to start contributing code.

**Comment by sipa** [comment]
@onefire Reasonable nits, but no reason to hold this up, IMHO. Feel free to submit improvements.

**Comment by paveljanik** [comment]
@onefire While doing so, please fix also the nits I brought. Thank you.

**Pull request merged**